[APIv3] Specify first iteration of WorkPackage PATCH-Endpoint

NobodysNightmare opened the pull request:
This PR belongs to [#16770](https://community.openproject.org/work_packages/16770).

It defines the very first portion of the `PATCH` endpoint for work packages.

NobodysNightmare committed:
fix aglio-warnings

NobodysNightmare committed:
specify expected format for PATCH endpoint

NobodysNightmare committed:
unify documentation of error responses

NobodysNightmare committed:
specify conflict cases

NobodysNightmare committed:
specify 403

NobodysNightmare committed:
improve example

TeatroIO commented:
I've prepared a stage. <a href='http://feature-specify-wp-patch.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

NobodysNightmare committed:
specify error objects

NobodysNightmare committed:
sort and improve error id list

NobodysNightmare committed:
shrink intro on yet unspecified responses

also clarifying the meaning of collections

NobodysNightmare committed:
explicitly ban html from error messages

NobodysNightmare committed:
extend error specification

- allow collection of errors
- define ValidationError

NobodysNightmare committed:
remove definition of HTTP status codes

this list does not add more value than the standard definitions...
it is not even nearly the list of values our API will return by any
specification

hschink commented on doc/apiv3-documentation.apib:
```diff
@@ -26,118 +26,90 @@ Read more: http://stateless.co/hal_specification.html
+* `urn:openproject-org:api:v3:errors:MissingPermission` (**HTTP 403**) - The client does not have the needed permissions to perform the requested action
+* `urn:openproject-org:api:v3:errors:NotFound` (**HTTP 404**) - Default for HTTP 404 when no further information is available
+* `urn:openproject-org:api:v3:errors:UpdateConflict` (**HTTP 409**) - The resource changed between GET-ing it and performing an update on it
+* `urn:openproject-org:api:v3:errors:PropertyIsReadOnly` (**HTTP 422**) - The client tried to set or change a property that is not writable
```
@NobodysNightmare I spoke with @myabc about the API's behavior and according to him it should be OK if the API just ignores read-only attributes. This would render this identifier obsolete. What do you think?

NobodysNightmare replied:
I think that I would not be okay with that.

Assume a client wants to set a property like `id` or `createdAt`. Those are not writable now and will never be. A client trying to set them obviously does not know this.

I therefore vote for returning an error. Because the request was not **OK** (HTTP 200), it was _bad_ and we could not (fully) fulfill it. And requests that can't be fulfilled completely should not be fulfilled at all, because this could lead to inconsistent state (assuming the client really meant to change a read-only field).

NobodysNightmare committed:
change notation of multiple errors

embed them in a single error object to ease implementation of simple
clients, advanced clients wont care anyway

NobodysNightmare committed:
more clarification on embedded error infos

NobodysNightmare commented:
I changed how the representation of multiple errors works after talking with @jpassing and @ulferts.

The basic motivation is that a client can always expect that he either receives the specified response body or a **single error**. This means a simple client only has to take into account those two cases to make a request and show a proper error message.

For the (few) cases where more than one error is needed, we use error nesting.

@myabc @hschink (and others): Tell me if you think the current specification (on errors and WP-Patch) is good or if there are points left for discussion.

hschink commented on doc/apiv3-documentation.apib:
```diff
@@ -26,118 +26,90 @@ Read more: http://stateless.co/hal_specification.html
+    
+#### List of Error Identifiers
+
+* `urn:openproject-org:api:v3:errors:MissingPermission` (**HTTP 403**) - The client does not have the needed permissions to perform the requested action
```
`401` is missing

hschink commented on doc/apiv3-documentation.apib:
```diff
@@ -26,118 +26,147 @@ Read more: http://stateless.co/hal_specification.html
+    
+## List of Error Identifiers
+
+* `urn:openproject-org:api:v3:errors:MissingPermission` (**HTTP 403**) - The client does not have the needed permissions to perform the requested action
```
`401` is missing

NobodysNightmare committed:
add HTTP 401 to MissingPermission identifier

NobodysNightmare committed:
be more precise about subject constraints

NobodysNightmare committed:
Merge remote-tracking branch 'origin/dev' into feature/specify_wp_patch

Conflicts:
	doc/apiv3-documentation.apib

NobodysNightmare committed:
remove WRITE operation where not implemented

write operation aws specified in many places that will not receive a
write-operation soon...
it will eventually be specified, but until then I'd rather have the
specs more consistent than scattering WRITE-operations throughout the
docs

hschink commented on doc/apiv3-documentation.apib:
```diff
@@ -1392,13 +1423,46 @@ Updates an activity's comment and, on success, returns the updated activity.
+                "_type": "ValidationError",
+                "errorIdentifier": "urn:openproject-org:api:v3:errors:PropertyConstraintViolation",
+                "message": "The subject might not be blank.",
+                "field": "Subject"
```
The current implementation looks like:

``` JavaScript
"_embedded": {
  "details": {
    "attribute": "Subject"
  }
}
```

I guess this matches the requirements stated above.

NobodysNightmare replied:
Right, my example violated the spec...

fixed

hschink commented on doc/apiv3-documentation.apib:
```diff
@@ -1379,12 +1397,25 @@ Updates an activity's comment and, on success, returns the updated activity.
+* The `_type` which should as of now always be `WorkPackage`
+* The value of `lockVersion` which was received by the `GET` request this change originates from.
+
+The value of `lockVersion` is used to implement [optimistic locking](http://en.wikipedia.org/wiki/Optimistic_concurrency_control).
```
This is planned but not implemented yet.

NobodysNightmare replied:
I'd just not mention that... From a client perspective we are complete here:
- the lockversion can be read
- the client sends the lockversion
- he always receives HTTP 200, but is prepared for HTTP 409

That the server will not currently fire a 409 is sad, but probably should not be mentioned in the specs (as opposed to places where the client would break if it relied on the spec).

hschink commented on doc/apiv3-documentation.apib:
```diff
@@ -26,118 +26,150 @@ Read more: http://stateless.co/hal_specification.html
+        "message": "This is an example error.",
+        "_embedded": {
+            "details": {
+                "_type": "ExampleErrorDetailInformation",
```
I hope this [is optional](https://tools.ietf.org/html/draft-kelly-json-hal-06#section-5.3)?

NobodysNightmare replied:
I did only mention it in an example so it is currently optional... However note, that this is different from the `type` you linked...

hschink commented on doc/apiv3-documentation.apib:
```diff
@@ -1392,13 +1423,46 @@ Updates an activity's comment and, on success, returns the updated activity.
+                "message": "You are not allowed to edit the content of the work package."
+            }
+    
++ Response 409 (application/hal+json)
```
As I stated above, this is not implemented at the moment

hschink commented on doc/apiv3-documentation.apib:
```diff
@@ -1392,13 +1423,46 @@ Updates an activity's comment and, on success, returns the updated activity.
+            {
+                "_type": "Error",
+                "errorIdentifier": "urn:openproject-org:api:v3:errors:MissingPermission",
+                "message": "You are not allowed to edit the content of the work package."
```
`You are not authorized to access this resource.` We don't state the actual resource.

NobodysNightmare replied:
This is an example and I hope nobody reads the message text in an example as part of the specification (as in: If the message is different it is a violation).
I'd rather keep this message as an example, because I'd really wish this _was_ the actual error message.

Also: I'd go nuts if I had to change every error message when the implementation for that changes.

NobodysNightmare committed:
make example spec-conform

also meets current implementation
